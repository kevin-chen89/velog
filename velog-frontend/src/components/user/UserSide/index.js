// @flow
export { default } from './UserSide';