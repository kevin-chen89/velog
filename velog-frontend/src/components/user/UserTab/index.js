// @flow
export { default } from './UserTab';