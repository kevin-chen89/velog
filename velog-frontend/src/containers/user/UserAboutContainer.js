// @flow
import React, { Component, Fragment } from 'react';
import { ProfileActions, BaseActions } from 'store/actionCreators';
import { type State } from 'store';
import { connect } from 'react-redux';
import UserAbout from 'components/user/UserAbout';
import UserAboutEdit from 'components/user/UserAboutEdit';
import { Helmet } from 'react-helmet';

type Props = {
  about: ?string,
  self: boolean,
  username: ?string,
  displayName: ?string,
};

type OwnProps = {
  username: ?string,
};

type UserAboutContainerState = {
  editing: boolean,
  text: string,
};

class UserAboutContainer extends Component<Props, UserAboutContainerState> {
  state = {
    editing: false,
    text: '',
  };

  onEditClick = () => {
    this.setState({
      editing: true,
    });
  };

  onSave = async () => {
    const { text } = this.state;
    try {
      await ProfileActions.updateAbout(text);
      BaseActions.showToast({
        message: '자기소개가 업데이트되었습니다.',
        type: 'success',
      });
    } catch (e) {
      BaseActions.showToast({
        message: '자기소개 업데이트 실패',
        type: 'error',
      });
    }
    this.setState({
      editing: false,
    });
  };

  onChange = (text: string) => {
    this.setState({
      text,
    });
  };

  componentDidMount() {
    ProfileActions.setSideVisibility(false);
  }

  componentWillUnmount() {
    ProfileActions.setSideVisibility(true);
  }

  render() {
    const { username, displayName } = this.props;
    if ((!this.props.about && this.props.about !== '') || !username || !displayName) return null;
    if (this.state.editing) {
      return <UserAboutEdit text={this.state.text} onChange={this.onChange} onSave={this.onSave} />;
    }
    return (
      <Fragment>
        {
          <Helmet>
            <title>{`About ${username} (${displayName}) | velog`}</title>
            <meta
              name="description"
              content={
                this.props.about === ''
                  ? `${displayName}님의 자기소개가 비어있습니다.`
                  : this.props.about
              }
            />
          </Helmet>
        }
        <UserAbout about={this.props.about} self={this.props.self} onEditClick={this.onEditClick} />
      </Fragment>
    );
  }
}

export default connect(({ profile, user }: State, ownProps: OwnProps) => ({
  username: profile.profile && profile.profile.username,
  displayName: profile.profile && profile.profile.display_name,
  about: profile.profile && profile.profile.about,
  self: (user.user && user.user.username) === ownProps.username,
}))(UserAboutContainer);
