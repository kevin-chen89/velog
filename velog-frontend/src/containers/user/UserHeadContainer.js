// @flow
import React, { Component } from 'react';
import * as actions from 'store/actionCreators';
import type { State } from 'store';
import { connect } from 'react-redux';
import UserHead from 'components/user/UserHead';

type Props = {};

class UserHeadContainer extends Component<Props> {
  render() {
    return <UserHead />;
  }
}

export default connect((state: State) => ({}), () => ({}))(UserHeadContainer);
