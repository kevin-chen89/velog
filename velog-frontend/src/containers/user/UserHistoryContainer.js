// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { ProfileActions } from 'store/actionCreators';
import UserHistory from 'components/user/UserHistory/UserHistory';
import { withRouter, type ContextRouter } from 'react-router-dom';
import { compose } from 'redux';
import { type State } from 'store';
import { type UserHistoryItem } from 'store/modules/profile';

type Props = {
  userHistory: ?(UserHistoryItem[]),
  prefetched: ?(UserHistoryItem[]),
  prefetching: boolean,
  loading: boolean,
} & ContextRouter;
class UserHistoryContainer extends Component<Props> {
  lastOffset: ?number;
  initialize = async () => {
    const { match } = this.props;
    const { username } = match.params;
    try {
      if (!username) return;
      await ProfileActions.getUserHistory({ username });
    } catch (e) {
      console.log(e);
    }
  };

  prefetch = async () => {
    const { userHistory, prefetched, loading, prefetching, match } = this.props;
    const { username } = match.params;
    if (!username || !userHistory || loading || prefetching) return;
    // REVEAL
    ProfileActions.revealPrefetchedHistory();
    await Promise.resolve();
    try {
      await ProfileActions.prefetchUserHistory({
        username,
        offset: userHistory.length || 0,
      });
    } catch (e) {
      console.log(e);
    }
  };

  componentDidMount() {
    ProfileActions.setSideVisibility(false);
    this.initialize();
  }
  componentWillUnmount() {
    ProfileActions.setSideVisibility(true);
  }

  render() {
    const { userHistory, match } = this.props;
    if (!userHistory) return null;
    return <UserHistory data={userHistory} username={match.params.username || ''} />;
  }
}

const enhance = compose(
  withRouter,
  connect(
    ({ profile, pender }: State) => ({
      userHistory: profile.userHistory,
      prefetched: profile.prefetchedHistory,
      loading: pender.pending['profile/GET_USER_HISTORY'],
      prefetching: pender.pending['profile/PREFETCH_USER_HISTORY'],
    }),
    () => ({}),
  ),
);

export default enhance(UserHistoryContainer);
