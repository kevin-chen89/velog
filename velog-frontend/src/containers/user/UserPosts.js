// @flow
import React, { Component } from 'react';
import PostCardList from 'components/common/PostCardList/PostCardList';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { withRouter } from 'react-router-dom';
import { ListingActions, ProfileActions } from 'store/actionCreators';
import type { State } from 'store';
import type { PostItem } from 'store/modules/listing';
import throttle from 'lodash/throttle';
import { getScrollBottom, preventStickBottom } from 'lib/common';

type OwnProps = {
  username: string,
  tag?: string,
};

type Props = OwnProps & {
  posts: ?(PostItem[]),
  prefetched: ?(PostItem[]),
  hasEnded: boolean,
  prefetching: boolean,
  loading: boolean,
  rawTagName: ?string,
};

class UserPosts extends Component<Props> {
  prevCursor: ?string = null;

  initialize = async () => {
    const { username, tag } = this.props;
    ListingActions.clearUserPosts();
    if (tag) {
      try {
        await ProfileActions.getTagInfo(tag);
      } catch (e) {
        console.log(e);
      }
    }
    const { rawTagName } = this.props;
    ListingActions.getUserPosts({ username, tag: rawTagName || undefined });
  };
  componentDidMount() {
    this.initialize();
    this.listenScroll();
  }
  componentWillUnmount() {
    this.unlistenScroll();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.tag !== this.props.tag || prevProps.username !== this.props.username) {
      this.initialize();
    }
  }

  prefetch = async () => {
    const { prefetched, hasEnded, posts, prefetching, loading, username, tag } = this.props;
    if (prefetched) {
      ListingActions.revealPrefetched('user');
      await Promise.resolve(); // next tick
    }
    if (hasEnded || prefetching || loading || !posts || posts.length === 0) return;
    const lastId = posts[posts.length - 1].id;
    if (this.prevCursor === lastId) return;
    this.prevCursor = lastId;
    try {
      await ListingActions.prefetchUserPosts({
        username,
        tag,
        cursor: lastId,
      });
      preventStickBottom();
    } catch (e) {
      console.log(e);
    }
  };

  onScroll = throttle(() => {
    const scrollBottom = getScrollBottom();
    if (scrollBottom > 1000) return;
    this.prefetch();
  }, 100);

  listenScroll = () => {
    window.addEventListener('scroll', this.onScroll);
  };

  unlistenScroll = () => {
    window.removeEventListener('scroll', this.onScroll);
  };

  render() {
    const { posts, loading, prefetching } = this.props;
    return (
      <PostCardList
        posts={posts}
        oneColumn
        loading={loading}
        prefetching={prefetching}
        hasEnded={false}
      />
    );
  }
}

export default connect(
  ({ listing, pender, profile }: State) => ({
    posts: listing.user.posts,
    prefetched: listing.user.prefetched,
    hasEnded: listing.user.end,
    prefetching: pender.pending['listing/PREFETCH_USER_POSTS'],
    loading: pender.pending['listing/GET_USER_POSTS'],
    rawTagName: profile.rawTagName,
  }),
  () => ({}),
)(UserPosts);
