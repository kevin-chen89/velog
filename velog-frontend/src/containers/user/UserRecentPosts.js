// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { withRouter, type Match } from 'react-router-dom';

type Props = {};
class UserRecentPosts extends Component<Props> {
  render() {
    return <div>배고파요</div>;
  }
}

export default compose(connect(), withRouter)(UserRecentPosts);
