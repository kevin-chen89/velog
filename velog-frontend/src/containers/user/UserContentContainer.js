// @flow
import React, { Component } from 'react';
import UserContent from 'components/user/UserContent';
import UserTagView from 'components/user/UserTagView';
import UserTab from 'components/user/UserTab';
import { withRouter, type Match, Route } from 'react-router-dom';
import { ProfileActions } from 'store/actionCreators';
import { type TagCountInfo, type Profile } from 'store/modules/profile';
import { connect } from 'react-redux';
import type { State } from 'store';
import { compose } from 'redux';
import UserPostsSubpage from 'pages/user/UserPostsSubpage';

type Props = {
  match: Match,
  tagCounts: ?(TagCountInfo[]),
  shouldCancel: boolean,
  profile: ?Profile,
};

class UserContentContainer extends Component<Props> {
  initialize = async () => {
    const { shouldCancel } = this.props;
    if (shouldCancel) return;
    const { username } = this.props.match.params;
    if (!username) return;
    if (username === (this.props.profile && this.props.profile.username)) return;
    ProfileActions.initialize();
    await ProfileActions.getUserTags(username);
    await ProfileActions.getProfile(username);
  };

  componentDidMount() {
    this.initialize();
  }

  onSelectTag = (tagName: string) => {
    ProfileActions.setRawTagName(tagName);
  };

  componentDidUpdate(prevProps) {
    if (this.props.match.params.username !== prevProps.match.params.username) {
      this.initialize();
    }
  }

  render() {
    const { match, tagCounts } = this.props;
    const username = match.params.username || '';

    return (
      <UserContent
        side={
          <UserTagView tagCounts={tagCounts} username={username} onSelectTag={this.onSelectTag} />
        }
      >
        <UserTab username={username}>
          <Route exact path="/@:username" component={UserPostsSubpage} />
          <Route path="/@:username/tags/:tag" component={UserPostsSubpage} />
        </UserTab>
      </UserContent>
    );
  }
}

export default compose(
  withRouter,
  connect(
    ({ profile, common }: State) => ({
      profile: profile.profile,
      tagCounts: profile.tagCounts,
      shouldCancel: common.ssr && !common.router.altered,
    }),
    () => ({}),
  ),
)(UserContentContainer);
