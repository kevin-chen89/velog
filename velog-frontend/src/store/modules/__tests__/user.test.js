// @flow
import user, { actionCreators, type User, UserActionCreators } from '../user';

describe('user', () => {
  describe('reducer', () => {
    let state: User = user((undefined: any), ({}: any));

    it('initializes state', () => {
      expect(state).toMatchSnapshot();
    });
    // TODO: test checkUser
    it('action: SET_USER', () => {
      const sampleUser = {
        id: 'id',
        username: 'velopert',
        displayName: 'velopert',
        thumbnail: 'asdf',
      };
      state = user(state, actionCreators.setUser(sampleUser));
      expect(state.user).toBe(sampleUser);
    });
    it('action: PROCESS_USER', () => {
      state = user(state, actionCreators.processUser());
      expect(state.processed).toBeTruthy();
    });
  });
});