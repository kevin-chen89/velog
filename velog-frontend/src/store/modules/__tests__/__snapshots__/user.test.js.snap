// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user reducer initializes state 1`] = `
Object {
  "processed": false,
  "user": null,
}
`;
